### PR TITLE
Add basic nn.Linear + nn.Tanh MLPs.

### PR DESCRIPTION
**Pull request by silvasean**
Adds a 1 and 2 layer MLP and run them E2E on RefBackend (IREE is still waiting for https://github.com/google/iree/issues/5409 to be resolved).

The main things needed here are:
- aten::linear op vertical slice
- InlineGlobalSlots pass to inline parameters.

The final IR  (+ canonicalization) for Mlp1Layer is:
https://gist.github.com/silvasean/e2b1899609bb3f3259fb7dc5235db664
It's kind of hard to see because we don't have the right mix of custom named ops for aten::linear to come out super pretty, but it's basically just pure linalg ops, with the minimal set of `dim` ops / error cases (some even folded away magically). (See the relevant commit changing ATenToLinalg for some details for the limitations here -- automating the aten to linalg route is going to be a nontrivial work item; optional args, variadic rank, element dtypes/type-promotion, size-1 broadcasting...).

Recommended review order: commit-by-commit
Note: stacked on top of: https://github.com/llvm/mlir-npcomp/pull/208, please ignore the other commits

### CODE REVIEW

File: lib/Dialect/Torch/Transforms/RefineTypes.cpp [hunk 1]
```diff
@@ -202,6 +202,21 @@ forwardKnowledgeTransferFunction(Operation *op,
     // the same, then the result is of that same element type.
     knowledge.elementType = joinElementTypes(lhs.elementType, rhs.elementType);
     return {knowledge};
+  } else if (isa<aten::LinearOp>(op)) {
+    // The output shape is the input shape with the last dimension changed
+    // to the weight's output dimension.
+    auto knowledge = operandKnowledge[0];
+    if (knowledge.hasRank && knowledge.sizes.size() > 0)
+      knowledge.sizes[knowledge.sizes.size() - 1] = kUnknownSize;
+    // TODO: Handle case of bias being None gracefully. Requires a lattice
```
Comment (stellaraccident) [inline lib/Dialect/Torch/Transforms/RefineTypes.cpp:211]: Wdyt of adding an issue category for starter issues? It seems like this would be a good one to note there.

**Review by stellaraccident — APPROVED** [review]
Looks like this included a diffbase (I wish the PR workflow was better) but looked past that. LGTM!

File: lib/Dialect/Torch/Transforms/RefineTypes.cpp (hunk 1 above)
Comment (silvasean) [reply lib/Dialect/Torch/Transforms/RefineTypes.cpp:211]: Filed https://github.com/llvm/mlir-npcomp/issues/211

I'm hoping to roll this into the new "aten-to-linalg + shape inference + aten/prim td generation + named op generation" component handles.
